chain/namada/namada_chain.go

```diff
@@ -0,0 +1,1291 @@
+	cmd := []string{
+		"namadac",
+		"--base-dir",
+		c.HomeDir(),
+		"transfer",
+		"--source",
+		keyName,
+		"--target",
+		amount.Address,
+		"--token",
+		amount.Denom,
+		"--amount",
+		amount.Amount.String(),
+		"--gas-payer",
+		gasPayerAlias,
+		"--node",
+		c.GetRPCAddress(),
+	}
+	_, _, err := c.Exec(ctx, cmd, c.Config().Env)
+
+	return err
+}
+
+// Generate an IBC shielding transfer for the following shielding transfer via IBC.
+func (c *NamadaChain) GenIbcShieldingTransfer(ctx context.Context, channelID string, amount ibc.WalletAmount, options ibc.TransferOptions) (string, error) {
+	var portID string
+	if options.Port == "" {
+		portID = "transfer"
+	} else {
+		portID = options.Port
+	}
+
+	cmd := []string{
+		"namadac",
+		"--base-dir",
+		c.HomeDir(),
+		"ibc-gen-shielding",
+		"--output-folder-path",
+		c.HomeDir(),
+		"--target",
+		amount.Address,
+		"--token",
+		amount.Denom,
+		"--amount",
+		amount.Amount.String(),
+		"--port-id",
+		portID,
+		"--channel-id",
+		channelID,
+		"--node",
+		c.GetRPCAddress(),
+	}
+	output, _, err := c.Exec(ctx, cmd, c.Config().Env)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate the IBC shielding transfer: %v", err)
+	}
+	outputStr := string(output)
+
+	re := regexp.MustCompile(`Output IBC shielding transfer for ([^\s]+) to (.+)`)
+	matches := re.FindStringSubmatch(outputStr)
+	var path string
+	if len(matches) > 2 {
+		path = matches[2]
+	} else {
+		return "", fmt.Errorf("failed to get the file path of the IBC shielding transfer")
+	}
+	relPath, _ := filepath.Rel(c.HomeDir(), path)
+	shieldingTransfer, err := c.Validators[0].ReadFile(ctx, relPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to read the IBC shielding transfer file: %v", err)
+	}
+
+	return string(shieldingTransfer), nil
+}
+
+// Get the current block height.
+func (c *NamadaChain) Height(ctx context.Context) (int64, error) {
+	return c.Validators[0].Height(ctx)
+}
+
+// Get the balance with the key alias, not the address.
+func (c *NamadaChain) GetBalance(ctx context.Context, keyName string, denom string) (math.Int, error) {
+	if strings.HasPrefix(keyName, "shielded") {
+		cmd := []string{
+			"namadac",
+			"--base-dir",
+			c.HomeDir(),
+			"shielded-sync",
+			"--viewing-keys",
+			keyName,
+			"--node",
+			c.GetRPCAddress(),
+		}
+		output, _, err := c.Exec(ctx, cmd, c.Config().Env)
+		if err != nil {
+			return math.NewInt(0), fmt.Errorf("shielded-sync failed: error %s, output %s", err, output)
+		}
+	}
+
+	cmd := []string{
+		"namadac",
+		"--base-dir",
+		c.HomeDir(),
+		"balance",
+		"--token",
+		denom,
+		"--owner",
+		keyName,
+		"--node",
+		c.GetRPCAddress(),
+	}
+	output, _, err := c.Exec(ctx, cmd, c.Config().Env)
+	if err != nil {
+		return math.NewInt(0), fmt.Errorf("getting the balance failed: error %s, output %s", err, output)
+	}
+	outputStr := string(output)
+	lines := strings.Split(outputStr, "\n")
+	re := regexp.MustCompile(`:\s*(\d+(\.\d+)?)$`)
```

Comment:
a comment explaining what this is / examples of matching content

Reply:
Added